applications/StructuralMechanicsApplication/custom_response_functions/adjoint_elements/adjoint_finite_difference_base_element.cpp

```diff
@@ -0,0 +1,603 @@
+            {
+                // disturb the design variable
+                node_i.GetInitialPosition()[coord_dir_i] += delta;
+
+                this->AfterPerturbation(rDesignVariable, rCurrentProcessInfo);
+
+                // compute RHS after disturbance
+                mpPrimalElement->CalculateRightHandSide(RHS_dist, copy_process_info);
+
+                //compute derivative of RHS w.r.t. design variable with finite differences
+                noalias(RHS_dist) -= RHS_undist;
+                RHS_dist /= delta;
+                for(unsigned int i = 0; i < RHS_dist.size(); i++)
+                    rOutput( (coord_dir_i + index*dimension), i) = RHS_dist[i];
+
+                // Reset perturbed vector
+                noalias(RHS_dist) = ZeroVector(RHS_dist.size());
+
+                // undisturb the design variable
+                node_i.GetInitialPosition()[coord_dir_i] -= delta;
+
+                this->AfterPerturbation(rDesignVariable, rCurrentProcessInfo);
+            }
+            index++;
+
+            //call one last time to make sure everything is as it was before TODO improve this..
+            mpPrimalElement->CalculateRightHandSide(RHS_dist, copy_process_info);
+
+        }// end loop over element nodes
+
+    }
+    else
+        KRATOS_ERROR << "Unsupported design variable!" << std::endl;
+
+    KRATOS_CATCH("")
+}
+
+void AdjointFiniteDifferencingBaseElement::CalculateStressDisplacementDerivative(const Variable<Vector>& rStressVariable,
+                                            Matrix& rOutput, const ProcessInfo& rCurrentProcessInfo)
+{
+    KRATOS_TRY;
+
+    const int num_nodes = mpPrimalElement->GetGeometry().PointsNumber();
+    const int dimension = mpPrimalElement->GetGeometry().WorkingSpaceDimension();
+    const int num_dofs_per_node = dimension * 2; // TODO generalize
+    const int num_dofs = num_nodes * num_dofs_per_node;
+    Vector initial_state_variables;
+    Vector stress_derivatives_vector;
+
+    // TODO first calculation only to get the size of the stress vector
+    this->Calculate(rStressVariable, stress_derivatives_vector, rCurrentProcessInfo);
+    rOutput.resize(num_dofs, stress_derivatives_vector.size() );
+    rOutput.clear();
+    initial_state_variables.resize(num_dofs);
+
+    // Build vector of variables containing the DOF-variables of the primal problem
+    std::vector<VariableComponent<VectorComponentAdaptor<array_1d<double, 3>>>> primal_solution_variable_list;
+    primal_solution_variable_list.reserve(num_dofs_per_node);
+    primal_solution_variable_list.push_back(DISPLACEMENT_X);
+    primal_solution_variable_list.push_back(DISPLACEMENT_Y);
+    primal_solution_variable_list.push_back(DISPLACEMENT_Z);
+    primal_solution_variable_list.push_back(ROTATION_X);
+    primal_solution_variable_list.push_back(ROTATION_Y);
+    primal_solution_variable_list.push_back(ROTATION_Z);
+
+    // TODO Armin
+    KRATOS_ERROR_IF(rCurrentProcessInfo.Has(NL_ITERATION_NUMBER))
```

Comment:
This is not the best way to check it

Dont have a better solution atm though ...

Reply:
Me neither, kept like it was